backend/drm/drm.c

```diff
@@ -1613,3 +1624,141 @@ void destroy_drm_connector(struct wlr_drm_connector *conn) {
 	wl_list_remove(&conn->link);
 	free(conn);
 }
+
+int wlr_drm_backend_get_read_only_fd(struct wlr_backend *backend) {
+	assert(backend);
+
+	struct wlr_drm_backend *drm = get_drm_backend_from_backend(backend);
+	char *path = drmGetDeviceNameFromFd2(drm->fd);
+	if (!path) {
+		wlr_log(WLR_ERROR, "Failed to get device name from DRM fd");
+		return -1;
+	}
+
+	int fd = open(path, O_RDWR);
+	if (fd < 0) {
+		wlr_log_errno(WLR_ERROR, "Unable to clone DRM fd for client fd");
+		free(path);
+		return -1;
+	}
+
+	if (drmIsMaster(fd) && drmDropMaster(fd) < 0) {
```

Comment:
I don't think this is enough? Tried running this little program from a TTY: https://l.sr.ht/jdqo.c

It can restore DRM master after dropping it.

I think we should just fail if `drmIsMaster` returns true.

Reply:
The ability should be restricted to the process that opened the FD. Try forking in your example.

See https://elixir.bootlin.com/linux/v5.12.9/source/drivers/gpu/drm/drm_auth.c#L204 for more details.

Reply:
Indeed, can't `drmSetMaster` in a child process.

@bl4ckb0ne, can you add a note to the function docs that the returned DRM FD is not master, but the compositor process must not call `drmSetMaster` on the returned FD?

Reply:
Done in the header https://github.com/swaywm/wlroots/pull/2929/files#diff-cd7bf513dc856e2f39ba4e5a6eb7e32de3e189fd25481f24fb30c1894b8b6c28R37